### ISSUE TITLE
Interpretting language mode from the files content

### ISSUE THREAD

**Issue by yashpalgoyal1304** [microsoft/vscode#136822]
* It would be nice to have some way to declare the language mode for the files which dont announce it in extension, and have vscode to interpret it  
i.e. `.txt` or extensionless files for example
* think of it as extension of `<!DOCTYPE html>` but for declaring other file types too
* i know that this syntax is not smth that vscode alone should decide, maybe W3C then? where can i go regarding that?
* or do some general syntax like this already exists??

**P.S.:** when i incurred this idea some days ago, i originally thought to discard it. but today, seeing #136263 , i thought it will be worth it to file it anyways

---
* **update1:** refer to this and following comments for some clarification: https://github.com/microsoft/vscode/issues/136822#issuecomment-968857321

**Comment by TylerLeonhardt** [comment]
@yashpalgoyal1304 so today we do automatic language detection based on the contents of the file:
https://code.visualstudio.com/updates/v1_60#_automatic-language-detection

@JacksonKearl is currently improving this support with a more accurate ML model...

Is this what you're looking for?

Or are you looking for an extension of:
![image](https://user-images.githubusercontent.com/2644648/141523921-6028c998-8053-4418-92fc-244a698f8997.png)
for arbitrary text/regex in the file?

We also already do some basic checking of the shebang for extensionless files. For example, create a file called `foo` with no extension... and let the body of the file be:

```
#!python

```

now close the file and reopen... it should show up as a python file now.

**Comment by yashpalgoyal1304** [comment]
1. » _automatic language detection_  
nope, i am not saying for that

2. » _Files: Associations_   
referring the documentation ([adding-extension-to-language]), did u meant associating any txt file to some particular language ( similar to what this asked for: #128985 )? if you mean so, then no, this isnt what i said. but it's nice to know about it anyways. 

[adding-extension-to-language]: https://code.visualstudio.com/docs/languages/overview#_adding-a-file-extension-to-a-language
[Known language identifiers]: https://code.visualstudio.com/docs/languages/identifiers#_known-language-identifiers

3. » _shebang_ (interpretter specifier)   
yeah, i meant smth like this, and it worked when used verbatim `#!python` . however, when i used `#!markdown` or `#!html` it didnt work. following from above referred documentation article to the "[Known language identifiers]", these should have worked.

**Comment by yashpalgoyal1304** [comment]
hi @TylerLeonhardt can u please remove the need more label lest the issue may get auto closed 😅

**Comment by TylerLeonhardt** [comment]
So the ask is, some kind of meta language/syntax inside of a file that is used to tell vscode what kind of file it is?

@bpasero this is more your territory, isn't it?

**Comment by bpasero** [comment]
Not really. We support that today already with the `firstLine` property, e.g.:

https://github.com/microsoft/vscode/blob/d465e988f1efe4c0d82ef92d633d38322cb5e8ca/extensions/make/package.json#L32

**Comment by TylerLeonhardt** [comment]
@bpasero is there a way to arbitrarily configure that? Like could a user do it without an extension?

**Comment by bpasero** [comment]
I don't think so. @alexdima might know.

**Comment by alexdima** [comment]
Extensions can use the API `languages.setTextDocumentLanguage` to change the language of a text document.

**Comment by yashpalgoyal1304** [comment]
@TylerLeonhardt u didnt remove the need more info label do u? 😅😅😅

> https://github.com/microsoft/vscode/issues/136822#issuecomment-970197419
